File: tests/benchmarks/control_flow_benchmark.jl
#!/usr/bin/env julia
# Control Flow Benchmark for Cross-Language Performance Testing
# Equivalent to control_flow_benchmark.orus

function main()
    start_time = time()
    
    # Test 1: Simple For Loop Performance (1 million iterations)
    simple_counter = 0
    for i in 0:999999
        simple_counter += 1
    end
    
    println(simple_counter)
    
    # Test 2: Nested Loop Performance (1000 x 1000)
    nested_total = 0
    for i in 0:999
        for j in 0:999
            nested_total += 1
        end
    end
    
    println(nested_total)
    
    # Test 3: While Loop with Conditional (100K iterations)
    while_counter = 0
    condition_hits = 0
    while while_counter < 100000
        if while_counter % 2 == 0
            condition_hits += 1
        end
        while_counter += 1
    end
    
    println(condition_hits)
    
    # Test 4: Conditional Logic (50K iterations)
    complex_result = 0
    for i in 0:49999
        if i % 3 == 0
            complex_result += 3
        else
            if i % 5 == 0
                complex_result += 5
            else
                complex_result += 1
            end
        end
    end
    
    println(complex_result)
    
    # Test 5: Loop with Conditional Processing (10K iterations)
    break_continue_total = 0
    processed_count = 0
    for i in 0:9999
        if i % 100 == 0
            break_continue_total += 0
        else
            break_continue_total += 1
            processed_count += 1
        end
    end
    
    println(break_continue_total)
    println(processed_count)
    
    # Test 6: Short Jump Stress Test - Tight Nested Loops
    tight_nested_total = 0
    for a in 0:199
        for b in 0:199
            for c in 0:4
                tight_nested_total += 1
            end
        end
    end
    
    println(tight_nested_total)
    
    # Test 7: Dense Conditionals
    dense_conditional_total = 0
    for i in 0:19999
        if i % 2 == 0
            dense_conditional_total += 1
        end
        if i % 3 == 0
            dense_conditional_total += 2
        end
        if i % 5 == 0
            dense_conditional_total += 3
        end
        if i % 7 == 0
            dense_conditional_total += 4
        end
    end
    
    println(dense_conditional_total)
    
    # Test 8: Mixed Control Flow
    mixed_total = 0
    for outer in 0:99
        inner_count = 0
        while inner_count < 50
            if inner_count % 3 == 0
                if outer % 2 == 0
                    mixed_total += 1
                else
                    mixed_total += 2
                end
            else
                mixed_total += 1
            end
            inner_count += 1
        end
    end
    
    println(mixed_total)
    
    end_time = time()
    println(stderr, "Julia execution time: $(end_time - start_time:.6f) seconds")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: tests/benchmarks/control_flow_benchmark.js
#!/usr/bin/env node
// Control Flow Performance Benchmark - JavaScript (For-Loop Version)
// Focus on for loops and optimized control flow patterns

console.log("=== JavaScript Control Flow Performance Benchmark ===");

// Constants for testing
const base1 = 12;
const base2 = 25;
const base3 = 37;
const multiplier = 7;
const offset = 100;
const factor = 3;

let total_result = 0;

// Test 1: Simple invariant expressions
console.log("Test 1: Simple invariant expressions");
for (let outer = 1; outer < 500; outer++) {
    for (let i = 1; i < 10; i++) {
        const expensive_calc1 = base1 * multiplier * 2;
        const expensive_calc2 = base2 + multiplier + offset;
        const expensive_calc3 = base3 * base1 + base2;
        
        const loop_var = i * factor;
        
        const result = expensive_calc1 + expensive_calc2 + expensive_calc3 + loop_var;
        total_result = total_result + result;
    }
}

// Test 2: Nested loops with complex expressions
console.log("Test 2: Nested loops with complex invariants");
for (let outer = 1; outer < 200; outer++) {
    for (let middle = 1; middle < 5; middle++) {
        for (let inner = 1; inner < 4; inner++) {
            const complex_calc1 = (base1 + base2) * (multiplier + offset);
            const complex_calc2 = base3 * base1 * base2 + multiplier;
            const complex_calc3 = (base1 * 2) + (base2 * 3) + (base3 * 4);
            
            const loop_dependent = inner * middle + outer;
            
            const result = complex_calc1 + complex_calc2 + complex_calc3 + loop_dependent;
            total_result = total_result + result;
        }
    }
}

// Test 3: Mixed invariant and variant expressions
console.log("Test 3: Mixed invariant and variant expressions");
for (let i = 1; i < 2000; i++) {
    const invariant1 = base1 * multiplier;
    const invariant2 = base2 + offset;
    const invariant3 = base3 * factor;
    
    const variant1 = i * 2;
    const variant2 = i + 10;
    
    const result = invariant1 + invariant2 + invariant3 + variant1 + variant2;
    total_result = total_result + result;
}

// Test 4: Conditional blocks with invariants
console.log("Test 4: Conditional blocks with invariants");
for (let i = 1; i < 1000; i++) {
    let result;
    if (i % 2 === 0) {
        const invariant_in_condition = base1 * base2 * multiplier;
        const variant_in_condition = i * 3;
        result = invariant_in_condition + variant_in_condition;
    } else {
        const other_invariant = base3 + offset + multiplier;
        const other_variant = i * 5;
        result = other_invariant + other_variant;
    }
    
    total_result = total_result + result;
}

console.log("Test 5: Function call simulation with invariants");
for (let i = 1; i < 800; i++) {
    const expensive_operation1 = base1 * base2 * base3;
    const expensive_operation2 = (multiplier + offset) * factor;
    const expensive_operation3 = base1 + base2 + base3 + multiplier + offset;
    
    const simple_operation = i + 1;
    
    const result = expensive_operation1 + expensive_operation2 + expensive_operation3 + simple_operation;
    total_result = total_result + result;
}

console.log("Control flow benchmark completed");
console.log("Total result:", total_result);
console.log("=== JavaScript Control Flow Benchmark Complete ===");